File: presto-native-execution/presto_cpp/main/functions/theta_sketch/ThetaSketchFunctions.cpp
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "presto_cpp/main/functions/theta_sketch/ThetaSketchRegistration.h"

#include "DataSketches/theta_sketch.hpp"

#include "velox/velox/functions/Macros.h"
#include "velox/velox/functions/Registerer.h"

namespace facebook::presto::functions {

namespace {

template <typename T>
struct ThetaSketchEstimateFunction {
  VELOX_DEFINE_FUNCTION_TYPES(T);

  FOLLY_ALWAYS_INLINE void call(
      out_type<double>& result,
      const arg_type<velox::Varbinary>& in) {
    auto compactSketch =
        datasketches::wrapped_compact_theta_sketch::wrap(in.data(), in.size());
    result = compactSketch.get_estimate();
  }
};

template <typename T>
struct ThetaSketchSummaryFunction {
  VELOX_DEFINE_FUNCTION_TYPES(T);
  FOLLY_ALWAYS_INLINE void call(
      out_type<velox::Row<double, double, double, double, int32_t>>& result,
      const arg_type<velox::Varbinary>& in) {
    auto compactSketch =
        datasketches::wrapped_compact_theta_sketch::wrap(in.data(), in.size());
    result.copy_from(
        std::make_tuple(
            compactSketch.get_estimate(),
            compactSketch.get_theta(),
            compactSketch.get_upper_bound(1),
            compactSketch.get_lower_bound(1),
            compactSketch.get_num_retained()));
  }
};
} // namespace

void registerThetaSketchFunctions(const std::string& prefix) {
  velox::
      registerFunction<ThetaSketchEstimateFunction, double, velox::Varbinary>(
          {prefix + "sketch_theta_estimate"});
  velox::registerFunction<
      ThetaSketchSummaryFunction,
      velox::Row<double, double, double, double, int32_t>,
      velox::Varbinary>({prefix + "sketch_theta_summary"});
}
} // namespace facebook::presto::functions

File: presto-main-base/src/main/java/com/facebook/presto/operator/scalar/ThetaSketchFunctions.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.facebook.presto.operator.scalar;

import com.facebook.presto.common.block.Block;
import com.facebook.presto.common.block.BlockBuilder;
import com.facebook.presto.common.type.RowType;
import com.facebook.presto.spi.function.Description;
import com.facebook.presto.spi.function.ScalarFunction;
import com.facebook.presto.spi.function.SqlType;
import com.google.common.collect.ImmutableList;
import io.airlift.slice.Slice;
import org.apache.datasketches.memory.Memory;
import org.apache.datasketches.theta.CompactSketch;

import java.nio.ByteOrder;
import java.util.Optional;

import static com.facebook.presto.common.type.DoubleType.DOUBLE;
import static com.facebook.presto.common.type.IntegerType.INTEGER;

public class ThetaSketchFunctions
{
    private ThetaSketchFunctions()
    {
    }

    @ScalarFunction(value = "sketch_theta_estimate")
    @Description("Get the estimate of unique values from a theta sketch")
    @SqlType("double")
    public static double thetaSketchEstimate(@SqlType("varbinary") Slice input)
    {
        CompactSketch sketch = CompactSketch.wrap(Memory.wrap(input.toByteBuffer(), ByteOrder.nativeOrder()));
        return sketch.getEstimate();
    }

    private static final RowType SUMMARY_TYPE = RowType.from(ImmutableList.of(
                new RowType.Field(Optional.of("estimate"), DOUBLE),
                new RowType.Field(Optional.of("theta"), DOUBLE),
                new RowType.Field(Optional.of("upper_bound_std"), DOUBLE),
                new RowType.Field(Optional.of("lower_bound_std"), DOUBLE),
                new RowType.Field(Optional.of("retained_entries"), INTEGER)));

    @ScalarFunction(value = "sketch_theta_summary")
    @Description("parses a brief summary from a theta sketch")
    @SqlType("row(estimate double, theta double, upper_bound_std double, lower_bound_std double, retained_entries int)")
    public static Block thetaSketchSummary(@SqlType("varbinary") Slice input)
    {
        CompactSketch sketch = CompactSketch.wrap(Memory.wrap(input.toByteBuffer(), ByteOrder.nativeOrder()));
        BlockBuilder output = SUMMARY_TYPE.createBlockBuilder(null, 1);
        BlockBuilder row = output.beginBlockEntry();
        DOUBLE.writeDouble(row, sketch.getEstimate());
        DOUBLE.writeDouble(row, sketch.getTheta());
        DOUBLE.writeDouble(row, sketch.getUpperBound(1));
        DOUBLE.writeDouble(row, sketch.getLowerBound(1));
        INTEGER.writeLong(row, sketch.getRetainedEntries());
        output.closeEntry();
        return output.build().getBlock(0);
    }
}
